Subject: Add rtstatic=yes option to sip.conf for RT/realtime without host=dynamic
From: Walter Doekes <wjdoekes@osso.nl>
Forwarded: no
Last-Update: 2017-03-17

This patch does two things:

 * It adds the rtstatic=yes option to sip.conf which will not cause the
   SIP realtime (ARA) to not query host=dynamic, because you're telling
   it that all hosts are statically configured.

   You'll only get name=FROMUSER queries.

   You will probably want to leave rtupdate=yes because if it does write
   anything after all, it won't end up in a local DB where nobody could
   see it.

 * Additionally, it makes the queue ringinuse work properly by setting
   the state_interface device to UNKNOWN instead of UNAVAILABLE.

   See also: https://reviewboard.asterisk.org/r/2996/
   See also: vg_uncached-peers-as-queuemembers.patch

--- a/channels/chan_sip.c
+++ b/channels/chan_sip.c
@@ -810,7 +810,7 @@ static char default_zone[MAX_TONEZONE_CO
 static unsigned int default_transports;            /*!< Default Transports (enum ast_transport) that are acceptable */
 static unsigned int default_primary_transport;     /*!< Default primary Transport (enum ast_transport) for outbound connections to devices */
 
-static struct sip_settings sip_cfg;		/*!< SIP configuration data.
+struct sip_settings sip_cfg;		/*!< SIP configuration data.
 					\note in the future we could have multiple of these (per domain, per device group etc) */
 
 /*!< use this macro when ast_uri_decode is dependent on pedantic checking to be on. */
@@ -5393,6 +5393,9 @@ static void update_peer(struct sip_peer
 	int rtcachefriends = ast_test_flag(&p->flags[1], SIP_PAGE2_RTCACHEFRIENDS);
 	if (sip_cfg.peer_rtupdate && (p->is_realtime || rtcachefriends)) {
 		struct ast_str *r = sip_route_list(&p->path, 0, 0);
+		if (sip_cfg.peer_rtstatic) {
+			ast_log(LOG_WARNING, "Unexpected update_peer for %s when rtstatic=yes\n", p->name);
+		}
 		if (r) {
 			realtime_update_peer(p->name, &p->addr, p->username,
 				p->fullcontact, p->useragent, expire, p->deprecated_username,
@@ -5518,7 +5521,8 @@ static const char *get_name_from_variabl
 static int realtime_peer_by_name(const char *const *name, struct ast_sockaddr *addr, const char *ipaddr, struct ast_variable **var, struct ast_variable **varregs)
 {
 	/* Peer by name and host=dynamic */
-	if ((*var = ast_load_realtime("sippeers", "name", *name, "host", "dynamic", SENTINEL))) {
+	/* host=dynamic lookup is skipped if we expect only static peers (rtstatic) */
+	if (!sip_cfg.peer_rtstatic && (*var = ast_load_realtime("sippeers", "name", *name, "host", "dynamic", SENTINEL))) {
 		;
 	/* Peer by name and host=IP */
 	} else if (addr && !(*var = ast_load_realtime("sippeers", "name", *name, "host", ipaddr, SENTINEL))) {
@@ -5595,6 +5599,12 @@ static int realtime_peer_by_addr(const c
 	/* We're not finding this peer by this name anymore. Reset it. */
 	*name = NULL;
 
+	/* Don't expect such a lookup for rtstatic. */
+	if (sip_cfg.peer_rtstatic) {
+		ast_log(LOG_WARNING, "Unexpected realtime_peer_by_addr for %s/%s/%s when rtstatic=yes\n",
+				*name, ipaddr, portstring);
+	}
+
 	/* First check for fixed IP hosts with matching callbackextensions, if specified */
 	if (!ast_strlen_zero(callbackexten) && (*var = ast_load_realtime("sippeers", "host", ipaddr, "port", portstring, "callbackextension", callbackexten, SENTINEL))) {
 		;
@@ -5817,6 +5827,7 @@ static struct sip_peer *sip_find_peer_fu
 		}
 	}
 
+	/* sip_devicestate() calls this with realtime=FALSE and devstate_only=TRUE */
 	if (!p && (realtime || devstate_only)) {
 		/* realtime_peer will return a peer with matching callbackexten if possible, otherwise one matching
 		 * without the callbackexten */
@@ -16627,6 +16638,9 @@ static void destroy_association(struct s
 
 	if (!sip_cfg.ignore_regexpire) {
 		if (peer->rt_fromcontact && sip_cfg.peer_rtupdate) {
+			if (sip_cfg.peer_rtstatic) {
+				ast_log(LOG_WARNING, "Unexpected destroy_association for %s when rtstatic=yes\n", peer->name);
+			}
 			ast_update_realtime(tablename, "name", peer->name, "fullcontact", "", "ipaddr", "", "port", "0", "regseconds", "0", "regserver", "", "useragent", "", "lastms", "0", SENTINEL);
 		} else {
 			ast_db_del("SIP/Registry", peer->name);
@@ -19812,7 +19826,7 @@ static void receive_message(struct sip_p
 		 * it to the right context.
 		 */
 
-		peer = sip_find_peer(NULL, &p->recv, TRUE, FINDPEERS, 0, p->socket.type);
+		peer = sip_find_peer(NULL, &p->recv, TRUE, FINDPEERS, FALSE, p->socket.type);
 		if (peer) {
 			/* Only if no auth is required. */
 			if (ast_strlen_zero(peer->secret) && ast_strlen_zero(peer->md5secret)) {
@@ -22081,6 +22095,7 @@ static char *sip_show_settings(struct as
 		ast_cli(a->fd, "  Realtime Regs:          %s\n", AST_CLI_YESNO(realtimeregs));
 		ast_cli(a->fd, "  Cache Friends:          %s\n", AST_CLI_YESNO(ast_test_flag(&global_flags[1], SIP_PAGE2_RTCACHEFRIENDS)));
 		ast_cli(a->fd, "  Update:                 %s\n", AST_CLI_YESNO(sip_cfg.peer_rtupdate));
+		ast_cli(a->fd, "  Realtime static only:   %s\n", AST_CLI_YESNO(sip_cfg.peer_rtstatic));
 		ast_cli(a->fd, "  Ignore Reg. Expire:     %s\n", AST_CLI_YESNO(sip_cfg.ignore_regexpire));
 		ast_cli(a->fd, "  Save sys. name:         %s\n", AST_CLI_YESNO(sip_cfg.rtsave_sysname));
 		ast_cli(a->fd, "  Save path header:       %s\n", AST_CLI_YESNO(sip_cfg.rtsave_path));
@@ -25010,6 +25025,9 @@ static void handle_response_peerpoke(str
 			peer->name, s, pingtime, peer->maxms);
 		ast_devstate_changed(AST_DEVICE_UNKNOWN, AST_DEVSTATE_CACHABLE, "SIP/%s", peer->name);
 		if (sip_cfg.peer_rtupdate) {
+			if (sip_cfg.peer_rtstatic) {
+				ast_log(LOG_WARNING, "Unexpected handle_response_peerpoke for %s when rtstatic=yes\n", peer->name);
+			}
 			ast_update_realtime(ast_check_realtime("sipregs") ? "sipregs" : "sippeers", "name", peer->name, "lastms", str_lastms, SENTINEL);
 		}
 		if (peer->endpoint) {
@@ -27932,7 +27950,7 @@ static int sip_msg_send(const struct ast
 	}
 
 	if (!ast_strlen_zero(from)) {
-		if ((peer_ptr = sip_find_peer(from, NULL, 0, 1, 0, 0))) {
+		if ((peer_ptr = sip_find_peer(from, NULL, FALSE, FINDUSERS, FALSE, 0))) {
 			ast_string_field_set(pvt, fromname, S_OR(peer_ptr->cid_name, peer_ptr->name));
 			ast_string_field_set(pvt, fromuser, S_OR(peer_ptr->cid_num, peer_ptr->name));
 			sip_unref_peer(peer_ptr, "sip_unref_peer, from sip_msg_send, sip_find_peer");
@@ -30736,6 +30754,7 @@ static int sip_devicestate(const char *d
 	char *host;
 	char *tmp;
 	struct sip_peer *p;
+	int devstate_only;
 
 	int res = AST_DEVICE_INVALID;
 
@@ -30752,8 +30771,19 @@ static int sip_devicestate(const char *d
 	 * (presumably to unavailable).  If we ask for a realtime peer here, this would
 	 * load it BACK into memory, thus defeating the point of trying to clear dead
 	 * hosts out of memory.
+	 *
+	 * Thus, we call for realtime=FALSE and devstate_only=TRUE.
+	 * Except when rtstatic=yes.  In that case we will skip the
+	 * realtime lookup altogether because even if we do find it
+	 * in the DB, we'll know two things: (a) that we didn't have
+	 * it in memory already (0 calls) and (b) that it exists
+	 * (NOT_INUSE instead of UNAVAILABLE).
+	 *
+	 * We'll short circuit that and return AST_DEVICE_NOT_INUSE
+	 * at once.
 	 */
-	if ((p = sip_find_peer(host, NULL, FALSE, FINDALLDEVICES, TRUE, 0))) {
+	devstate_only = !sip_cfg.peer_rtstatic;
+	if ((p = sip_find_peer(host, NULL, FALSE, FINDALLDEVICES, devstate_only, 0))) {
 		if (!(ast_sockaddr_isnull(&p->addr) && ast_sockaddr_isnull(&p->defaddr))) {
 			/* we have an address for the peer */
 
@@ -30788,11 +30818,18 @@ static int sip_devicestate(const char *d
 				res = AST_DEVICE_UNAVAILABLE;
 			else	/* Default reply if we're registered and have no other data */
 				res = AST_DEVICE_NOT_INUSE;
+		} else if (sip_cfg.peer_rtstatic && p->is_realtime) {
+			/* if this is a (uncached) realtime peer, we don't
+			 * necessarily have an address */
+			res = AST_DEVICE_UNKNOWN;
 		} else {
 			/* there is no address, it's unavailable */
 			res = AST_DEVICE_UNAVAILABLE;
 		}
 		sip_unref_peer(p, "sip_unref_peer, from sip_devicestate, release ref from sip_find_peer");
+	} else if (sip_cfg.peer_rtstatic) {
+		/* Static device and no other data */
+		res = AST_DEVICE_UNKNOWN;
 	}
 
 	return res;
@@ -32724,6 +32761,7 @@ static int reload_config(enum channelrel
 	ast_set_flag(&global_flags[1], SIP_PAGE2_ALLOWSUBSCRIBE);	/* Default for all devices: TRUE */
 	ast_set_flag(&global_flags[1], SIP_PAGE2_ALLOWOVERLAP_YES);	/* Default for all devices: Yes */
 	sip_cfg.peer_rtupdate = TRUE;
+	sip_cfg.peer_rtstatic = FALSE;
 	global_dynamic_exclude_static = 0;	/* Exclude static peers */
 	sip_cfg.tcp_enabled = FALSE;
 	sip_cfg.websocket_enabled = TRUE;
@@ -32855,6 +32893,8 @@ static int reload_config(enum channelrel
 			sip_cfg.rtsave_path = ast_true(v->value);
 		} else if (!strcasecmp(v->name, "rtupdate")) {
 			sip_cfg.peer_rtupdate = ast_true(v->value);
+		} else if (!strcasecmp(v->name, "rtstatic")) {
+			sip_cfg.peer_rtstatic = ast_true(v->value);
 		} else if (!strcasecmp(v->name, "ignoreregexpire")) {
 			sip_cfg.ignore_regexpire = ast_true(v->value);
 		} else if (!strcasecmp(v->name, "timert1")) {
@@ -35662,7 +35702,7 @@ static int load_module(void)
 	/* And start the monitor for the first time */
 	restart_monitor();
 
-	if (sip_cfg.peer_rtupdate) {
+	if (sip_cfg.peer_rtupdate && !sip_cfg.peer_rtstatic) {
 		ast_realtime_require_field(ast_check_realtime("sipregs") ? "sipregs" : "sippeers",
 			"name", RQ_CHAR, 10,
 			"ipaddr", RQ_CHAR, INET6_ADDRSTRLEN - 1,
--- a/channels/sip/include/sip.h
+++ b/channels/sip/include/sip.h
@@ -745,6 +745,7 @@ struct __show_chan_arg {
 */
 struct sip_settings {
 	int peer_rtupdate;          /*!< G: Update database with registration data for peer? */
+	int peer_rtstatic;          /*!< G: Only static entries found in DB? */
 	int rtsave_sysname;         /*!< G: Save system name at registration? */
 	int rtsave_path;            /*!< G: Save path header on registration */
 	int ignore_regexpire;       /*!< G: Ignore expiration of peer  */
@@ -792,6 +793,8 @@ struct sip_settings {
 
 struct ast_websocket;
 
+extern struct sip_settings sip_cfg;		/*!< SIP configuration data. A global. */
+
 /*! \brief The SIP socket definition */
 struct sip_socket {
 	enum ast_transport type;  /*!< UDP, TCP or TLS */
--- a/channels/sip/security_events.c
+++ b/channels/sip/security_events.c
@@ -272,6 +272,7 @@ int sip_report_security_event(const char
     const struct sip_request *req, const int res)
 {
 
+	int realtime;
 	struct sip_peer *peer_report;
 	enum check_auth_result res_report = res;
 	struct ast_str *buf;
@@ -288,7 +289,8 @@ int sip_report_security_event(const char
 		[K_LAST]  = { NULL, NULL}
 	};
 
-	peer_report = sip_find_peer(peer, addr, TRUE, FINDPEERS, FALSE, p->socket.type);
+	realtime = sip_cfg.peer_rtstatic ? FALSE : TRUE; /* no extra DB lookups for rtstatic, please */
+	peer_report = sip_find_peer(peer, addr, realtime, FINDPEERS, FALSE, p->socket.type);
 
 	switch(res_report) {
 	case AUTH_DONT_KNOW:
--- a/configs/samples/sip.conf.sample
+++ b/configs/samples/sip.conf.sample
@@ -1112,6 +1112,9 @@ srvlookup=yes                   ; Enable
                                 ; If not present, defaults to 'yes'. Note: realtime peers will
                                 ; probably not function across reloads in the way that you expect, if
                                 ; you turn this option off.
+;rtstatic=no                    ; SPECIAL HACK: Enable if all your realtime hosts are statically
+                                ; configured. No lookups will be done against host=dynamic and
+                                ; device state lookups will be skipped if not found in memory.
 ;rtautoclear=yes                ; Auto-Expire friends created on the fly on the same schedule
                                 ; as if it had just registered? (yes|no|<seconds>)
                                 ; If set to yes, when the registration expires, the friend will
